### PR TITLE
Add `ExtensionList.lookupFirst` convenience method.

### PR DESCRIPTION
**Pull request by Vlatombe**
When introducing an extension point that is meant to allow overriding a default behaviour with another implementation, generally the caller only cares about looking up only one implementation, the one with the highest ordinal.

See [JENKINS-XXXXX](https://issues.jenkins.io/browse/JENKINS-XXXXX).

### Proposed changelog entries

- developer: A new convenience method `ExtensionList.lookupFirst` allows to retrieve the first implementation of an extension point

### Proposed upgrade guidelines

N/A

```[tasklist]
### Submitter checklist
- [ ] The Jira issue, if it exists, is well-described.
- [X] The changelog entries and upgrade guidelines are appropriate for the audience affected by the change (users or developers, depending on the change) and are in the imperative mood (see [examples](https://github.com/jenkins-infra/jenkins.io/blob/master/content/_data/changelogs/weekly.yml)). Fill in the **Proposed upgrade guidelines** section only if there are breaking changes or changes that may require extra steps from users during upgrade.
- [ ] There is automated testing or an explanation as to why this change has no tests.
- [ ] New public classes, fields, and methods are annotated with `@Restricted` or have `@since TODO` Javadocs, as appropriate.
- [ ] New deprecations are annotated with `@Deprecated(since = "TODO")` or `@Deprecated(forRemoval = true, since = "TODO")`, if applicable.
- [ ] New or substantially changed JavaScript is not defined inline and does not call `eval` to ease future introduction of Content Security Policy (CSP) directives (see [documentation](https://www.jenkins.io/doc/developer/security/csp/)).
- [ ] For dependency updates, there are links to external changelogs and, if possible, full differentials.
- [ ] For new APIs and extension points, there is a link to at least one consumer.
```

### Desired reviewers

@mention

Before the changes are marked as `ready-for-merge`:

```[tasklist]
### Maintainer checklist
- [ ] There are at least two (2) approvals for the pull request and no outstanding requests for change.
- [ ] Conversations in the pull request are over, or it is explicit that a reviewer is not blocking the change.
- [ ] Changelog entries in the pull request title and/or **Proposed changelog entries** are accurate, human-readable, and in the imperative mood.
- [ ] Proper changelog labels are set so that the changelog can be generated automatically.
- [ ] If the change needs additional upgrade steps from users, the `upgrade-guide-needed` label is set and there is a **Proposed upgrade guidelines** section in the pull request title (see [example](https://github.com/jenkinsci/jenkins/pull/4387)).
- [ ] If it would make sense to backport the change to LTS, a Jira issue must exist, be a _Bug_ or _Improvement_, and be labeled as `lts-candidate` to be considered (see [query](https://issues.jenkins.io/issues/?filter=12146)).
```

### CODE REVIEW

File: core/src/main/java/hudson/ExtensionList.java
Comment (jglick) [inline core/src/main/java/hudson/ExtensionList.java:464]:
```suggestion
     * Equivalent to {@code ExtensionList.lookup(type).get(0)} if there is at least one instance,
```

File: core/src/main/java/hudson/ExtensionList.java
Comment (jglick) [inline core/src/main/java/hudson/ExtensionList.java:465]:
```suggestion
     * and throws an {@link IllegalStateException} otherwise if no instance could be found.
```

File: core/src/main/java/hudson/ExtensionList.java
Comment (jglick) [inline core/src/main/java/hudson/ExtensionList.java:469]:
```suggestion
     * @throws IllegalStateException if there are no instances
```

**Review by jglick — APPROVED** [review]
A couple of Javadoc nits, but otherwise looks useful. Did you search for other possible usages in core?

**Comment by Vlatombe** [comment]
Surprisingly, I didn't find other usages in core. I think `ArtifactManagerFactory` could be reworked to use this pattern, as there is a fallback mecanism to `StandardArtifactManager` if no artifact manager was configured for a given build.

**Comment by jglick** [comment]
https://github.com/jenkinsci/jenkins/blob/a23a8e58fb5894ec53db7ddb01cffbb17ce4828e/core/src/main/java/hudson/model/Run.java#L1127-L1135 uses a different pattern actually. So does https://github.com/jenkinsci/jenkins/blob/a23a8e58fb5894ec53db7ddb01cffbb17ce4828e/core/src/main/java/jenkins/console/DefaultConsoleUrlProvider.java#L44-L46

**Comment by jglick** [comment]
also checked https://github.com/jenkinsci/cloudbees-folder-plugin/blob/ed1b220693d41bb6e444d2274ffbae0684cc4a4b/src/main/java/com/cloudbees/hudson/plugins/folder/relocate/DefaultRelocationUI.java#L63-L72

**Review by timja — APPROVED** [review]
/label ready-for-merge

----

This PR is now ready for merge, after ~24 hours, we will merge it if there's no negative feedback.

Thanks!

**Pull request merged**
